### ISSUE TITLE
Best Effort Inventory Building

### ISSUE THREAD

**Issue by gbrindisi** [GoogleCloudPlatform/forseti-security#241]
Hello,
was wondering if it would make sense to make the inventory building more resilient.

For example sometimes I get errors when querying the APIs and an exception will be raised on the pipeline. The whole snapshot cycle is then labeled as `PARTIAL_SUCCESS` and the pipeline is considered done.

This is mildly annoying when operating on a big pipeline like the `LoadGroupMembersPipeline`: our rationale is that a partial inventory is better than an empty one.

A recent example:

```
2017-04-24 16:43:19,789 __main__     DEBUG    _run_pipelines Running pipeline <google.cloud.security.inventory.pipelines.load_group_members_pipeline.LoadGroupMembersPipeline object at 0x10a2d1490>
2017-04-24 16:44:48,776 __main__     ERROR    _run_pipelines Encountered error loading data.
GCP API Error: unable to get <googleapiclient.discovery.Resource object at 0x10a9cb590> from GCP:
<HttpError 503 when requesting https://www.googleapis.com/admin/directory/v1/groups/XYZ/members?alt=json&pageToken=XXX returned "Service unavailable. Please try again">
2017-04-24 16:44:48,776 __main__     INFO     _run_pipelines Continuing on.
2017-04-24 16:44:48,852 __main__     INFO     _complete_snapshot_cycle Inventory load cycle completed with PARTIAL_SUCCESS: 20170424T143633Z
```

The exception handling is in `inventory_loader.py`:

```
    for pipeline in pipelines:
        try:
            LOGGER.debug('Running pipeline %s', pipeline)
            pipeline.run()
            pipeline.status = 'SUCCESS'
        except inventory_errors.LoadDataPipelineError as e:
            LOGGER.error('Encountered error loading data.\n%s', e)
            pipeline.status = 'FAILURE'
            LOGGER.info('Continuing on.')
        run_statuses.append(pipeline.status == 'SUCCESS')
    return run_statuses
```

Could it be possible to collect the errors received, without raising an exception and try to keep going with the pipeline?
Eventually add the registered errors in a dedicated column within the `snapshot_cycles` table, for audit purposes.

For example:

```
+----+---------------------+---------------------+-----------------+----------------+------------------+------------------+
| id | start_time          | complete_time       | status          | schema_version | cycle_timestamp  | errors           |
+----+---------------------+---------------------+-----------------+----------------+------------------+------------------+
|  9 | 2017-04-24 14:36:33 | 2017-04-24 14:44:49 | PARTIAL_SUCCESS | 1.0            | 20170424T143633Z | JSON             |
+----+---------------------+---------------------+-----------------+----------------+------------------+------------------+
```

Where `JSON` might be something flexible like:

```
{
 "pipelines": {
  "load_group_members_pipeline": [
    {
        "group_id" : 123,
        "reason: "returned 503: Service unavailable."
    },
    ...
  ]
 }
}
```

And to avoid getting blocked on too many errors we might use a tolerance cap to decide whether proceed to the next pipeline (i.e. _keep going until you hit N errors_).

Thoughts?

**Comment by carise** [comment]
Great timing -- Henry (@blueandgold) and I were just chatting about this for a somewhat related subject (best effort loading data and how to store/surface the error). I think this would be a great improvement for the pipelines. Will let Henry comment.

**Comment by blueandgold** [comment]
Hi Gianluca, those are excellent points, and make great sense to us.  Since you have already done a lot of thinking here and even wrote some of the psuedo-code, would you like to send a PR to formalize this?  That would be super awesome!  :) 

A couple of pointers that would make the PR easier:

1: Every ```self._execute(request)``` call in the API code is already wrapped by a retry decorator as defined in the base API code.  Do you think this retry limit could serve as the tolerance cap?

[common/gcp_api/_base_client.py](https://github.com/GoogleCloudPlatform/forseti-security/blob/master/google/cloud/security/common/gcp_api/_base_client.py#L52)
```python
    @retry(retry_on_exception=retryable_exceptions.is_retryable_exception,
           wait_exponential_multiplier=1000, wait_exponential_max=10000,
           stop_max_attempt_number=5)
    def _execute(self, request):
```

2: However, the googleapiclient HttpError is not one of the retryable exceptions.  So, depending on the outcome of the 503 that we are debugging, this might need to be added as a retryable exception.  

[common/util/retryable_exceptions.py
](https://github.com/GoogleCloudPlatform/forseti-security/blob/master/google/cloud/security/common/util/retryable_exceptions.py#L25)
```python
RETRYABLE_EXCEPTIONS = (
    httplib.ResponseNotReady,
    httplib.IncompleteRead,
    httplib2.ServerNotFoundError,
    socket.error,
    ssl.SSLError,
    urllib2.URLError,  # no network connection is included here
)
```

**Comment by gbrindisi** [comment]
@blueandgold excellent, I will investigate a solution.
